### DISCUSSION TITLE
DAGs not running, cannot retrieve any logs.

### DISCUSSION THREAD

**Discussion by clementchane** [apache/airflow#31108] [General] [answered]
### Official Helm Chart version

1.9.0 (latest released)

### Apache Airflow version

2.5.3

### Kubernetes Version

Client Version: v1.26.0 Kustomize Version: v4.5.7 Server Version: v1.25.6

### Helm Chart configuration

Basic values file

fernetKey: ~
webserverSecretKey: ~

images:
  airflow:
   repository: privaterepo/airflow
   tag: v0.5

scheduler:
  livenessProbe:
    initialDelaySeconds: 30
    timeoutSeconds: 60
    failureThreshold: 5
    periodSeconds: 60

env:
  - name: "AIRFLOW_VAR_."
    value: "#{...}#"
registry:
  secretName: ...

### Docker Image customizations

FROM apache/airflow:2.5.3
USER root
RUN apt-get update \
  && apt-get install -y --no-install-recommends \
         vim \
  && apt-get autoremove -yqq --purge \
  && apt-get clean \
  && rm -rf /var/lib/apt/lists/*
USER airflow

RUN python -m pip install --upgrade pip \
  && pip install airflow-code-editor \
  && pip install black fs-s3fs fs-gcsfs requests PyGithub psycopg2 datetime cryptography==38.0.4 xlrd==1.2.0 pygithub jaydebeapi ndpmetadata

ENTRYPOINT [ "/usr/bin/dumb-init","--","/entrypoint" ]

### What happened

Installation seems to work fine, my DAGs, variables, and modules appear in the UI.
However I cannot run anything, and I have no logs.
I did not find any relevant logs in the webserver/scheduler pods
<img width="631" alt="image" src="https://user-images.githubusercontent.com/112370548/235575018-2cb26b39-d1d0-4f22-9694-66dec7d66008.png">

<img width="653" alt="image" src="https://user-images.githubusercontent.com/112370548/235575145-5f7bbbd8-7b61-45de-93ec-e1f8d30d636d.png">

### What you think should happen instead

At least have some logs on why DAGs failed.

### How to reproduce

Install apache-airflow/airflow v2.5.3 helm charts. 
Added a DAG directly in the webserver dags folder.

Dag is never running, I have reinstalled several times, but never managed to get logs in Airflow.

### Anything else

I have tried to look in all the closed issues or discussions, but hard to find any clear answers on why do I get no logs.
I have tried looking at pod logs and events but nothing helpful.

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by boring-cyborg** [comment]
Thanks for opening your first issue here! Be sure to follow the issue template! If you are willing to raise PR to address this issue please do so, no need to wait for approval.

**Comment by potiuk** [comment]
This is troubleshooting question - not an airflow issue. It should be moved to discussion, not created as an issue (we explain it in the explanatory comments when you open an issue).

Unfortunately - without logs, it is impossible to say what is wrong, but there are many more logs you can look for the reasons (not all of them even can be shown in the logs, because there might be certain reasons, that the problems you have (usually resource and configuration based) prevent from even starting task - which in turn will not make it possible to write to the logs. You should look in other logs:

* each component of Airflow has "logs" folder in AIRFLOW_HOME - you should be able to exec to running containers and check the logs there. for example you will find information in scheduler logs if it cannot actually run a task via whatever executor it has if the executor is badly configured.

* kubernetes also has its own logs where for example it can log information if there are no resources to run your tasks - for example when there is not enough memory to start it - `k9s` is a nice tool I often use to be able to see the logs and browse them - in case you are not familiar wiht k8s, this is also a way how to learn how to operate and diagnose k8s installation (which you will need anyway).

Generally - like most things in IT - there are variousl layers involved - application, deployment, containers, virtual machines, pod orchestrations etc. and when you run appication on top of those layers, often the applications cannot even know what the problem is because it's the lower layer that handles that (and does not even expose it to the layer below). So you need really learn and master operating and diagnosing errors on all the layers to be able to effectively manage such applications.  Airflow's task logs often won' be  simply able to display the reasons.

**Reply by clementchane** [reply]
Hello,

Yes you are right sorry for creating an issue!
I know it s a troubleshooting question, but since I am new to Airflow, and using almost every default configurations, I was hoping to get some startup checklists.

I have added this in the values file to allocate more resources, but didnt find any recommended values.
```
webserver:  
  resources:
    limits:
      cpu: "1"
      memory: "5Gi"
    requests:
      cpu: "1"
      memory: "5Gi"
  livenessProbe:
    initialDelaySeconds: 30
    timeoutSeconds: 100
    failureThreshold: 5
    periodSeconds: 600


workers:
  resources:
    limits:
      cpu: "1"
      memory: "4Gi"
    requests:
      cpu: "1"
      memory: "4Gi"
```

And I figured out that I had to run airflow db init, airflow scheduler and airflow celery worker commands to get it running.
Now I have a few tasks running in my DAGs, but it ends up failing in the worker pod for this reason:
```
File "/home/airflow/.local/lib/python3.7/site-packages/airflow/utils/cli.py", line 232, in get_dag
    f"Dag {dag_id!r} could not be found; either it does not exist or it failed to parse."
airflow.exceptions.AirflowException: Dag 'tutorialtest' could not be found; either it does not exist or it failed to parse.
[2023-05-04 00:10:07,207: ERROR/ForkPoolWorker-15] Task airflow.executors.celery_executor.execute_command[9c9af8fb-928c-4ae4-84fc-a263952eab51] raised unexpected: AirflowException('Celery command failed on host: airflow-worker-0.airflow-worker.airflow.svc.cluster.local with celery_task_id 9c9af8fb-928c-4ae4-84fc-a263952eab51')
```

I have found many similar issues on the net, but still havent any clear solution on how to fix it.
Do you have any idea?

Thanks a lot for your time and your help

**Reply by potiuk** [reply]
There are no recommended values, because there aren't any. Today's world of computing is not as simple as it was in 2000 - you have different clouds, machine types, virtual ones, kubernetes is managing resources, depending on where you run your workflows (local kubernetes? cloud kubernetes? Which cloud? Autoscaling? etc. Depending on which version of python, which dags, which version of Airlfow and which version (and what flavor of - Redshift, GKE, AKS)  kubernetes, your requirements might be very different. Kubernetes runs a lot of components inside - mostly networking, DNS, monitoring, often Prometheus, memory and pod management etc. etc. Kubernetes is there to run multiple apps, it can have multiple nodes with different kinds of networking and different kinds of persistent storage that can have different memory and CPU characteristics. 

The time when you could say - those are the requirements and checklist when you run a kubernetes application are gone long time ago. 

When you chose K8S as your deployment platform you chose the complexity and you chose to be able to deploy, configure monitor and adapt your application to match whatever all the variability above gives you. And there is simply no way to give you simple checklist. Running application has become a very complex process where instead of being able to predict everything and set the values anc checklist, you should know your tools and be able to observe and adapt your system when you need to do so. In process control theory there are two types of processes - there are predictable ones, when you set parameters and they give you defined output, and there are the more complex processes which have so many variables thet the only possible way to manage them is to observe and adapt in a continuous loop. Running apps on K8S is definitely the latter.

If you want much better predictability - we have this "Running airflow in Docker" https://airflow.apache.org/docs/apache-airflow/stable/howto/docker-compose/index.html - which is a great start and has some defined minimum requirements and great to play with (though if you want to make it production-ready, you will face similar issues as above with K8S, this is only a quick start that allows you to easily get running Airflow instance and run some basic dags. 

Coming back to your problem: you  have a problem with parsing your DAGs on the worker. It looks like for some reason your DAG cannot be parsed on the worker. Either you are not following best practices https://airflow.apache.org/docs/apache-airflow/stable/best-practices.html#top-level-python-code and you are doing something wrong at the top-level code of the worker, (for exampel importing something that is not available on your worker), or for some reason your DAGs are missing on the worker pods (all DAGs should be present in the same place (DAGS folder) see https://airflow.apache.org/docs/apache-airflow/stable/administration-and-deployment/modules_management.html#modules-management and be parseable there. 

There might be many root causes for that. For example when you have persistence configured but one of the problems with your K8S flavour is that persistence is not properly configured. Some local development K8S (there are multiple flaviours and it is impossible to give a recipe for that) require configuring persistence in order to make it works. Or the persistence might require some system level depdencies (for example some k8S instances use NFS for persistence and need NFS to be installed on the system level and properly configured - but this is something that you have to check with your k8S and K8S should produce the right log and warnings in the way it usually produces them (you need to check your K8S logs for that in the way that your K8S recommends it) in order to find out (This is the part of "monitor and adapt") . And there might be more problems - your filesystem migh have to need some permissions configuration for example (this is the case when the file is there but airflow user has no permission to read it). Those are just two reasons, but there are likely many others. 

I genereally - if you have not yet managed k8s and your k8s does not provide good monitoring tools - recommend `k9s` tool which allows you to explore and look around your k8s instance and running applications, logs etc. It's nice to peek at and learn k8s.

Eventually - what you need to end up with is that you should be able to parse (`python <DAGS__FOLDER>/your_dag_file.py` and it should succeed). You can exec into your worker pod and try it out.

**Reply by clementchane** [reply]
Hello,

Thanks a lot for taking the time to write such a detailed answer!
Everything you said makes sense, and helps me a lot.

Actually, I am creating DAGs by putting them directly in the webserver pod Dags folder (using the Airflow Code Editor).
Since I cannot put the DAGs in the worker pod directly (need root access), I was thinking it would take them directly in the webserver.
Actually it works for the 1st task of myDAG before throwing me this "dag not found" error in the worker logs.
As I am using everything as simple as possible in the default values.yaml, I didnt enable any persistency yet.

I will try setting that up and will let you know!
Anyway, thanks a lot for the time you are taking to help, that s great.

**Reply by clementchane** [reply]
Hello,

Thanks again for your time.
I forgot to say, but I m working on AKS and now added an Azure File Share. All the DAGs are correct since they were working on my VM.
Some progress have been made with the persistent storage, but now I have some random behaviors on the DAGs, they fail at some random tasks  with no config changes.
<img width="261" alt="image" src="https://user-images.githubusercontent.com/112370548/237028048-231f1c3e-9bab-490e-89b9-81568afc815d.png">

I monitor the pods, and it s not running out of memory or cpu.
When I look at the worker logs, sometimes it has logs on triggers, but sometimes it doesnt when the DAG fails, so I m guessing that some tasks dont reach the worker pod...
So basically, I have no logs and random behaviors, which is hard to debug :(

If you have any ideas, or things that I could test, I m open to any suggestions! Else I will keep making some other tests.
Thanks again!

**Reply by potiuk** [reply]
I recommend git-sync, init containers and our Helm Chart pretty much always when people have problems with shared storage. Don't know if this is your case but here is my article you might want to read: https://medium.com/apache-airflow/shared-volumes-in-airflow-the-good-the-bad-and-the-ugly-22e9f681afca

**Reply by clementchane** [reply]
Thanks! That s a great article.
Unfortunately, we cannot use git, as we want to provide airflow as a package to our customers with prebuilt DAGs and configs, so we cannot have use our own git, and we wont have access to the customers git :(

**Reply by clementchane** [reply, marked as answer]
It seems that the issue is that the scheduler sometimes cannot reach the worker pod.
No tasks are triggered in the worker, and the task is failing without any logs.
I have added some retries values inside the DAGs and now it runs until the end after some time, so it looks ok! (even if it can take up to 20mn instead of 2mn)
Thank you very much for your time and your help, your advices were very helpful.
